### PR TITLE
STYLE: Format code using `ruff`

### PR DESCRIPTION
**Pull request by jhlegarreta**
Format the code using `ruff`: supersedes `flake8` and `blue`.

`blue` was not being applied previously as the relevant `pre-commit` hook config block was commented in the.

Remove the `.flake8` config file: in practice few rules were being applied as the config file contained a significant number fo ignored rules.

Bump `ruff-pre-commit` pre-commit hook version to v0.4.3 (May 3, 2024): the version being used previously was v0.1.7 was from Dec 4, 2023. Made changes consistent between changes being observed locally vs on the GitHub action.

Sort the imports according to the new version: new sorting is dictated by the default value of `force-sort-within-sections` (`false`): imports are sorted by module.

### CODE REVIEW

**Comment by jhlegarreta** [comment]
The pre-commit hook is nicely doing its job:
https://github.com/fury-gl/fury/actions/runs/8861237757/job/24332897396?pr=881#step:4:23191

File: pyproject.toml [hunk 1]
```diff
@@ -97,19 +97,38 @@ raw-options = { version_scheme = "release-branch-semver" }
...
+]
+ignore = [
+  "E203"
+]
+
+[tool.ruff.lint.flake8-quotes]
+inline-quotes = "double"
+
+[tool.ruff.format]
+quote-style = "double"
```
Comment (skoudoro) [inline pyproject.toml:128]: I am not sure about this. 

double quotes vs single quotes. For now, we were using single quotes in this project.

File: .pep8speaks.yml [hunk 2]
```diff
@@ -1,12 +0,0 @@
-scanner:
```
Comment (skoudoro) [inline .pep8speaks.yml:1]: I know it is a duplicate but I I prefer to keep pep8speaks for now. 
people react more and faster on a new comment than a failed CI.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -16,15 +16,6 @@ repos:
...
       - id: detect-private-key
-  # - repo: https://github.com/grantjenks/blue
-  #   rev: v0.9.1
-  #   hooks:
-  #     - id: blue
-  - repo: https://github.com/pycqa/flake8
-    rev: 6.0.0
-    hooks:
-      - id: flake8
-        exclude: "^(docs/experimental|tools)/"
```
Comment (skoudoro) [inline .pre-commit-config.yaml:27]: where is it reflected in ruff? 

ok to lint `docs` and `tools` but we should avoid to lint `experimental`

File: pyproject.toml [hunk 4]
```diff
@@ -97,19 +97,38 @@ raw-options = { version_scheme = "release-branch-semver" }
...
 version-file = "fury/_version.py"
 
-[tool.blue]
+[tool.ruff]
 line_length = 88
 target-version = ["py38"]
-force-exclude = """
-(
-  _version.py
-  *docs/experimental/*
```
Comment (skoudoro) [inline pyproject.toml:106]: same comment as above

File: pyproject.toml [hunk 5]
```diff
@@ -97,19 +97,38 @@ raw-options = { version_scheme = "release-branch-semver" }
...
-
-[flake8]
-max-line-length = 88
-extend-exclude = ["_version.py", "externals", "*docs/experimental*"]
+exclude = [
+    ".git",
+    "__pycache__",
+    "build",
+    "dist",
+    "_version.py",
```
Comment (skoudoro) [inline pyproject.toml:108]: docs/experimental to add

**Review by skoudoro — CHANGES_REQUESTED** [review]
Hi @jhlegarreta,

Thank you for this. See my comment below

File: pyproject.toml (hunk 1 above)
Comment (jhlegarreta) [reply pyproject.toml:128]: `ruff` uses `double` by default, but I prefer to make that explicit.

Other projects:
- Nibabel: single: https://github.com/nipy/nibabel/blob/a2e5dee05cf374c22670ff9fd0d385ce366eb495/pyproject.toml#L141
- NumPy: does not enforce a quote style yet, contains both, e.g.: https://github.com/numpy/numpy/blob/970476737080a053d9be3d6f4479bb1b5cd17f19/tools/lint_diff.ini e.g. https://github.com/numpy/numpy/blob/970476737080a053d9be3d6f4479bb1b5cd17f19/numpy/ma/core.py
- SciPy does not enforce a quote style https://github.com/scipy/scipy/blob/main/tools/lint.toml, contains both e.g. https://github.com/scipy/scipy/blob/main/scipy/optimize/optimize.py
- Scikit-image: not sure, default `ruff` is double: https://github.com/scikit-image/scikit-image/blob/main/pyproject.toml#L154, but they have both in the code, e.g.
 https://github.com/scikit-image/scikit-image/blob/main/skimage/io/collection.py
https://github.com/scikit-image/scikit-image/blob/main/skimage/filters/_sparse.py
- Scikit-learn: not sure, default `ruff` is double: https://github.com/scikit-learn/scikit-learn/blob/main/pyproject.toml#L122 looks double quotes, e.g.
https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/compose/_column_transformer.py
https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/decomposition/_lda.py

File: .pre-commit-config.yaml (hunk 3 above)
Comment (jhlegarreta) [reply .pre-commit-config.yaml:27]: Here https://github.com/fury-gl/fury/pull/881/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R103

Not sure why experimental should not be linted. The rest of the exclusions apply to the build, dist, etc. folders that are not kept under version control, but IMO anything under version control should be linted.

File: pyproject.toml (hunk 4 above)
Comment (jhlegarreta) [reply pyproject.toml:106]: Same comment as above :point_up:.

File: pyproject.toml (hunk 5 above)
Comment (jhlegarreta) [reply pyproject.toml:108]: Same comment as above :point_up:.

File: .pep8speaks.yml (hunk 2 above)
Comment (jhlegarreta) [reply .pep8speaks.yml:1]: OK. Will keep it in the new push force, but we may run into conflicts if slight differences exist in the way `ruff` enforces some rules. We'll keep an eye on this.

File: pyproject.toml (hunk 1 above)
Comment (skoudoro) [reply pyproject.toml:128]: ok, let's go with double quotes

File: .pre-commit-config.yaml (hunk 3 above)
Comment (skoudoro) [reply .pre-commit-config.yaml:27]: experimental will be removed and we slowly do it step by step. It does not make any more sense to lint this folder. It contains deprecated code, bad formated code, etc. 
Everytime we move out something from experimental, we clean it to respect all the standard

File: pyproject.toml (hunk 4 above)
Comment (skoudoro) [reply pyproject.toml:106]: same comment as above

**Review by skoudoro — COMMENTED** [review]
Hi @jhlegarreta,

See above my comment. In summary:
- ok for double quotes
- remove experimental from linting please

**Comment by jhlegarreta** [comment]
@skoudoro OK. Please, have a look at the CI result to see if the changes look OK to you, and if so, I will apply them locally and commit all files.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (jhlegarreta) [reply .pre-commit-config.yaml:27]: OK. Adding it to the excluded file list then.

File: .pep8speaks.yml (hunk 2 above)
Comment (jhlegarreta) [reply .pep8speaks.yml]: Restored it in 20e1a17.

File: pyproject.toml (hunk 4 above)
Comment (jhlegarreta) [reply pyproject.toml:106]: Cross-referencing https://github.com/fury-gl/fury/pull/881#discussion_r1585605369. Excluding `experimental`. Resolving.

File: pyproject.toml (hunk 5 above)
Comment (jhlegarreta) [reply pyproject.toml:108]: Cross-referencing https://github.com/fury-gl/fury/pull/881#discussion_r1585605369. Excluding `docs/experimental`. Resolving.

**Comment by skoudoro** [comment]
Looking into it. I might merge https://github.com/fury-gl/fury/pull/885 and #883 before you do that.

**Comment by skoudoro** [comment]
I just skim it. A lot is single quotes vs double quotes. Not sure I will have time now to check if there is something weird that I do not agree. 

So for now, I agree and if something come up, it will be updated on a new PR.

**Comment by jhlegarreta** [comment]
> Looking into it. I might merge https://github.com/fury-gl/fury/pull/885 and https://github.com/fury-gl/fury/pull/883 before you do that.

> So for now, I agree and if something come up, it will be updated on a new PR.

OK. Will keep an eye on those PRs. If the issues in there are not addressed by Friday, I will lint the files as told by `ruff` here and will commit them.

**Comment by skoudoro** [comment]
@WassCodeur, please, see the comment above. You need to update all your PR before thursday afternoon

**Comment by WassCodeur** [comment]
sure

**Comment by jhlegarreta** [comment]
Some comments
- Add 2 exclusion rules (besides some paths)

- Running
   ```
   ruff --config pyproject.toml --fix
   ```

   detected some stuff that I had to undo so that tests pass, e.g.
   https://github.com/fury-gl/fury/blob/master/fury/ui/elements.py#L3232
   `warnings.warn` stacklevels and double quotes:
   https://github.com/fury-gl/fury/blob/master/fury/tests/test_testing.py#L105
   Writing as literals:
   https://github.com/fury-gl/fury/blob/master/fury/deprecator.py#L320
   https://github.com/fury-gl/fury/blob/master/docs/source/ext/github_tools.py#L412

  Probably my fixes were not good.

   Not sure why all the changes that the above command proposed were not being reported in the CI in the first place, or why the empty line before the first import is enforced only depending on how we call the commands (above vs `ruff format --diff`). 

   Looks like I'm missing something (GHA command outcome vs what I would expect, locally running the commands).

- No sure why the exclude rules
   https://github.com/fury-gl/fury/pull/881/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R108

   are not being applied:
https://github.com/fury-gl/fury/actions/runs/8946250187/job/24576577739#step:4:188

  Same goes to the `# codespell:ignore` in `fury/shaders.base.py`:
  https://github.com/fury-gl/fury/actions/runs/8946250187/job/24576577739#step:4:146

Anyways, I think the above differences can be worked in a separate PR: many of the edits in this PR were not trivial/were manual, and avoiding to go over all of them again would save time, e.g.




So I'd merge this as is and rework the rest in separate PRs. @skoudoro let me know if you can afford investigating the above a bit or whether merging this is better.

**Comment by skoudoro** [comment]
Thanks for this huge work.

Not sure to understand all your request but I will go carefully through this PR. if all ok, I will go ahead and merge, then, we will fix the other issue in new PRs

**Comment by jhlegarreta** [comment]
👍 One of the things that is maybe happening is that the `pyproject.toml` config options are not being raken into account for the commands above/GitHub format action, as when running . The action and the `.pre-commit-config` are pretty standard;

**Comment by jhlegarreta** [comment]
OK, got this working: the issue was that there is a `ruff.toml` file that I had completely forgotten about, and I was adding `ruff` rules to `pyproject.toml`, the rules int `ruff.toml` taking precedence. Tested that the pre-commit hook is raising all these on my fork:
https://github.com/jhlegarreta/fury/actions/runs/8953728977/job/24592533047?pr=3

And here fixes have been applied.

Sorry (esp @skoudoro) for all the noise generated by the push forces and the previous misleading comments.

This is ready to be merged (IMHO, the Ubuntu Python 3.9 failure is unrelated).

**Review by skoudoro — APPROVED** [review]
thank you @jhlegarreta,

merging

**Pull request merged**